core/metrics-core-service/src/main/java/org/hawkular/metrics/core/service/MetricsServiceImpl.java

```diff
@@ -655,7 +656,16 @@ public void setDefaultTTL(int defaultTTL) {
                         .doOnNext(i -> insertedDataPointEvents.onNext(metric)))
                 .doOnNext(meter::mark);
 
-        Observable<Integer> indexUpdates = dataAccess.updateMetricsIndex(metrics)
+        Observable<Metric<T>> updateIdxMetrics = metrics.filter(metric -> {
+            if (metricsIdxCache.getIfPresent(metric.getMetricId()) == null) {
+                metricsIdxCache.put(metric.getMetricId(), true);
```

Comment:
It does not make sense for anyone else reading the code that suddenly removing items actually writes items and changes the behavior elsewhere. This just makes the code difficult to follow. And there's no advantage to doing it here.

It's also not a good idea for backpressure aware processing, as filter should not be the one causing backpressure (and IIRC, filters do not support backpressure).